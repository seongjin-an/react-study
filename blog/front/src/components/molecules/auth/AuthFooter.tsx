import React from "react";
import styled from "styled-components";
import palette from "../../../libs/styles/palette";

export const AuthFooter: React.FC<{
    children: React.ReactNode
}> = ({children}) => {
    return<StyledAuthFooter>{children}</StyledAuthFooter>
}
export type AuthFooterProps = React.ComponentProps<typeof AuthFooter>
const StyledAuthFooter = styled.div`
  margin-top: 2rem;
  text-align: right;
  a {
    color: ${palette.gray[6]};
    text-decoration: underline;
    &:hover{
      color: ${palette.gray[9]};
    }
  }
`