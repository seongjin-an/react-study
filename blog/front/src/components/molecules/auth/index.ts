export type {AuthFormProps} from './AuthForm'
export {AuthForm} from './AuthForm'
export type {AuthFooterProps} from './AuthFooter'
export {AuthFooter} from './AuthFooter'