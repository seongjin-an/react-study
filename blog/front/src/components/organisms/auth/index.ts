export type {AuthAreaProps} from './AuthArea'
export {AuthArea} from './AuthArea'