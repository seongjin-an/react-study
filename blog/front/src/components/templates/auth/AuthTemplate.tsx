import React from "react";
import styled from "styled-components";
import palette from "../../../libs/styles/palette";
import { Link } from 'react-router-dom'

export const AuthTemplate: React.FC<{
    children: React.ReactNode
}> = ({ children }) => {
    return<StyledAuthTemplate>
        <StyledWhiteBox>
            <div className="logo-area">
                <Link to="/">REACTERS</Link>
            </div>
            {children}
        </StyledWhiteBox>
    </StyledAuthTemplate>
}
export type AuthTemplateProps = React.ComponentProps<typeof AuthTemplate>
const StyledAuthTemplate = styled.div`
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  right: 0;
  background: ${palette.gray[2]};
  display:flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`

const StyledWhiteBox = styled.div`
  .logo-area{
    display: block;
    padding-bottom: 2rem;
    text-align: center;
    font-weight: bold;
    letter-spacing: 2px;
  }
  box-shadow: 0 0 8px rgba(0, 0, 0, 0.25);
  padding: 2rem;
  width: 360px;
  background: white;
  border-radius: 2px;
`