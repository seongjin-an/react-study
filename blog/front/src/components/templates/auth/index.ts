export type {AuthTemplateProps} from './AuthTemplate'
export {AuthTemplate} from './AuthTemplate'