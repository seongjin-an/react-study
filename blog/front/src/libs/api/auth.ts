import client from './client'

export const login = ({username, password}:{username:string, password:string}) => {
    return client.post('/api/auth/login', {username, password})
}

export const register = ({username, password}:{username:string, password:string}) => {
    return client.post('/api/auth/register', {username, password})
}
export const check = () => client.get('/api/auth/check')

export const logout = () => client.post('/api/auth/logout')