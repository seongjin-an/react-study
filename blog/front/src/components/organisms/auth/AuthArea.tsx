import React, {ChangeEvent, MouseEvent, useEffect, useState} from "react";
import styled from "styled-components";
import palette from "../../../libs/styles/palette";
import {AuthForm} from "../../molecules/auth";
import {useDispatch, useSelector} from "react-redux";
import {RootState} from "../../../modules";
import {changeField, initializeForm, loginAction, registerAction} from "../../../modules/auth/authAction";
import {checkAction} from "../../../modules/user/userAction";
import { withRouter } from "react-router-dom";
import {check} from "../../../libs/api/auth";
export enum EFormType {
    login, register
}
export const AuthArea: React.FC<{
    type: EFormType
}> = ({type}) => {
    const dispatch = useDispatch()
    const [error, setError] = useState<string>('')
    const {form, auth, authError, user} = useSelector(({auth, user}: RootState) => ({
        form: type === EFormType.login ? auth.login : auth.register,
        auth: auth.auth,
        authError: auth.authError,
        user: user.user
    }))
    const handleChangeInput = (event:ChangeEvent<HTMLInputElement>):any => {
        const {name, value} = event.target
        dispatch(changeField({form: type, key: name, value: value}))
    }
    const handleSubmit = (event:MouseEvent<HTMLButtonElement>) => {
        event.preventDefault()
        if(type===EFormType.login){
            const {username, password} = form
            dispatch(loginAction({username, password}))
        }else if (type===EFormType.register){
            const {username, password, passwordConfirm} = form
            if([username, password, passwordConfirm].includes('')){
                setError('빈 칸을 모두 입력하세요.')
                return
            }
            if(password!==passwordConfirm){
                setError('비밀번호가 일치하지 않습니다.')
                return
            }
            dispatch(registerAction({username, password}))
        }
    }
    useEffect(() => {
        dispatch(initializeForm({form: type}))
    },[type])
    useEffect(() => {
        if(authError){
            console.log('오류 발생')
            console.log(authError)
            if(type===EFormType.login) {
                setError('로그인 실패')
                return
            }else if(type===EFormType.register){
                if(authError.response.status === 409) {
                    setError('이미 존재하는 계정명입니다.')
                    return
                }
                setError('회원가입 실패')
                return
            }
        }
        if(auth){
            if(type===EFormType.login) {
                console.log('로그인 성공')
                dispatch(checkAction())
            }else if(type===EFormType.register){
                console.log('회원가입 성공')
                console.log(auth)
                dispatch(checkAction())
            }
        }
    }, [auth, authError, dispatch])
    useEffect(() => {
        if(user){
            console.log('check api 성공')
            console.log('user')
            window.location.href='/'//react-router-dom withRouter() history.push('/')
        }
    }, [user])
    return<StyledAuthArea>
        <h3>{type===EFormType.login? '로그인' : '회원가입'}</h3>
        <AuthForm type={type} onChange={handleChangeInput} onSubmit={handleSubmit} error={error}/>
    </StyledAuthArea>
}

export type AuthAreaProps = React.ComponentProps<typeof AuthArea>
const StyledAuthArea = styled.div`
  h3{
    margin: 0;
    color: ${palette.gray[8]};
    margin-bottom: 1rem;
  }
`