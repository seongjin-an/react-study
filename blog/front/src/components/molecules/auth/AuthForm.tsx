import React, {ChangeEvent, MouseEvent} from "react";
import {Input} from "../../atoms/input";
import {AuthFooter} from "./AuthFooter";
import {Link} from 'react-router-dom'
import {CommonButton} from "../../atoms/button";
import {EFormType} from "../../organisms/auth/AuthArea";
import styled from "styled-components";
export const AuthForm: React.FC<{
    type: EFormType,
    onChange: (event:ChangeEvent<HTMLInputElement>) => {}
    onSubmit: (event:MouseEvent<HTMLButtonElement>) => void
    error: string
}> = ({type, onChange, onSubmit, error}) => {
    return<>
        <form>
            <Input autoComplete={"username"} name={"username"} placeholder={"아이디"} onChange={onChange}/>
            <Input type={"password"} autoComplete={"new-password"} name={"password"} placeholder={"비밀번호"} onChange={onChange}/>
            {type===EFormType.register&&
                <Input type={"password"} autoComplete={"new-password"} name={"passwordConfirm"} placeholder={"비밀번호 확인"} onChange={onChange}/>
            }
            {error && <ErrorMessage>{error}</ErrorMessage>}
            <CommonButton cyan fullWidth style={{marginTop: '1rem'}} onClick={onSubmit}>
                {type===EFormType.login? '로그인' : '회원가입'}
            </CommonButton>
        </form>
        <AuthFooter>
            {type===EFormType.login?
                <Link to="/register">회원가입</Link> :
                <Link to="/login">로그인</Link>
            }
        </AuthFooter>
    </>
}
export type AuthFormProps = React.ComponentProps<typeof AuthForm>

const ErrorMessage = styled.div`
  color: red;
  text-align: center;
  font-size: 0.875rem;
  margin-top: 1rem;
`