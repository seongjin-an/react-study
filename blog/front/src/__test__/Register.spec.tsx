import {Store} from "@reduxjs/toolkit";
import React from "react";
import {configure, mount} from "enzyme";
import {Provider} from "react-redux";
import {BrowserRouter} from "react-router-dom";
// import createMockStore from "redux-mock-store";
import configureMockStore from "redux-mock-store";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import * as redux from 'react-redux'
import {RegisterPage} from "../pages/register";
import toJson from "enzyme-to-json";
import {EFormType} from "../components/organisms/auth/AuthArea";


let store: Store
const setup = (WrappedComponent: React.ComponentProps<any>, props = {}) => {
    const wrapper = mount(
        <Provider store={store}>
            <BrowserRouter>
                <WrappedComponent {...props} />
            </BrowserRouter>
        </Provider>
    )
    return {
        wrapper
    }
}

describe('register component', () => {
    let spyOnUseSelector
    let spyOnUseDispatch
    let mockDispatch: jest.Mock
    let mockStore = configureMockStore()
    beforeEach(() => {
        store = mockStore()
        configure({adapter: new Adapter()})
        spyOnUseSelector = jest.spyOn(redux, 'useSelector')
        spyOnUseDispatch = jest.spyOn(redux, 'useDispatch')
        mockDispatch = jest.fn()
        spyOnUseSelector.mockImplementation(selectorFn => selectorFn({
            auth: {
                register: {
                    username: '',
                    password: '',
                    passwordConfirm: ''
                }
            },
            authError: null,
            user:{
                _id: '',
                username: ''
            }
        }))
        spyOnUseDispatch.mockReturnValue(mockDispatch)
    })
    it('should match snapshot', () => {
        const {wrapper} = setup(RegisterPage, {
            type: EFormType.register
        })

        expect(toJson(wrapper)).toMatchSnapshot()
    })
    it('should render properly', () => {
        const { wrapper } = setup(RegisterPage, {
            type: EFormType.register
        })
        const header = wrapper.find('.logo-area')
        expect(header.text()).toBe('REACTERS')
        const signup = wrapper.find('h3')
        expect(signup.text()).toBe('회원가입')
        const authInput = wrapper.find('AuthInput')
        expect(authInput).toHaveLength(3)
        const commonButton = wrapper.find('CommonButton')
        expect(commonButton.text()).toBe('회원가입')
        const footerButton = wrapper.find('AuthFooter')
        expect(footerButton.text()).toBe('로그인')
    })
    it('should fill', () => {
        const { wrapper } = setup(RegisterPage, {
            type: EFormType.register
        })
        const authInput = wrapper.find('AuthInput')
        authInput.at(0).simulate('change', { target: { name: 'username', value: 'imsi11' } })
        expect(mockDispatch.mock.calls[1][0]).toEqual({ type: 'auth/CHANGE_FIELD', payload: { form: 1, key: 'username', value: 'imsi11'}})
        authInput.at(1).simulate('change', { target: { name: 'password', value: '1234' } })
        expect(mockDispatch.mock.calls[2][0]).toEqual({ type: 'auth/CHANGE_FIELD', payload: { form: 1, key: 'password', value: '1234' }})
        authInput.at(2).simulate('change', { target: { name: 'passwordConfirm', value: '1234' } })
        expect(mockDispatch.mock.calls[3][0]).toEqual({ type: 'auth/CHANGE_FIELD', payload: { form: 1, key: 'passwordConfirm', value: '1234' }})
        console.log('STORE:', store.getState())
        // const usernameInput = wrapper.find('input[name="username"]')
        // console.log('usernameInput prop:', usernameInput.props())
    })
})