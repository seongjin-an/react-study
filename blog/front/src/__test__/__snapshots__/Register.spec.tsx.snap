// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register component should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <RegisterPage
        type={1}
      >
        <AuthTemplate>
          <styled.div>
            <div
              className="sc-bdfBQB binGTg"
            >
              <styled.div>
                <div
                  className="sc-gsTEea dahxES"
                >
                  <div
                    className="logo-area"
                  >
                    <Link
                      to="/"
                    >
                      <LinkAnchor
                        href="/"
                        navigate={[Function]}
                      >
                        <a
                          href="/"
                          onClick={[Function]}
                        >
                          REACTERS
                        </a>
                      </LinkAnchor>
                    </Link>
                  </div>
                  <withRouter(AuthArea)
                    type={1}
                  >
                    <AuthArea
                      history={
                        Object {
                          "action": "POP",
                          "block": [Function],
                          "createHref": [Function],
                          "go": [Function],
                          "goBack": [Function],
                          "goForward": [Function],
                          "length": 1,
                          "listen": [Function],
                          "location": Object {
                            "hash": "",
                            "pathname": "/",
                            "search": "",
                            "state": undefined,
                          },
                          "push": [Function],
                          "replace": [Function],
                        }
                      }
                      location={
                        Object {
                          "hash": "",
                          "pathname": "/",
                          "search": "",
                          "state": undefined,
                        }
                      }
                      match={
                        Object {
                          "isExact": true,
                          "params": Object {},
                          "path": "/",
                          "url": "/",
                        }
                      }
                      type={1}
                    >
                      <styled.div>
                        <div
                          className="sc-pGacB cfKvFf"
                        >
                          <h3>
                            회원가입
                          </h3>
                          <AuthForm
                            error=""
                            onChange={[Function]}
                            onSubmit={[Function]}
                            type={1}
                          >
                            <form>
                              <AuthInput
                                autoComplete="username"
                                name="username"
                                onChange={[Function]}
                                placeholder="아이디"
                                type="text"
                              >
                                <styled.input
                                  autoComplete="username"
                                  name="username"
                                  onChange={[Function]}
                                  placeholder="아이디"
                                  type="text"
                                >
                                  <input
                                    autoComplete="username"
                                    className="sc-dlfnuX hrdAtQ"
                                    name="username"
                                    onChange={[Function]}
                                    placeholder="아이디"
                                    type="text"
                                  />
                                </styled.input>
                              </AuthInput>
                              <AuthInput
                                autoComplete="new-password"
                                name="password"
                                onChange={[Function]}
                                placeholder="비밀번호"
                                type="password"
                              >
                                <styled.input
                                  autoComplete="new-password"
                                  name="password"
                                  onChange={[Function]}
                                  placeholder="비밀번호"
                                  type="password"
                                >
                                  <input
                                    autoComplete="new-password"
                                    className="sc-dlfnuX hrdAtQ"
                                    name="password"
                                    onChange={[Function]}
                                    placeholder="비밀번호"
                                    type="password"
                                  />
                                </styled.input>
                              </AuthInput>
                              <AuthInput
                                autoComplete="new-password"
                                name="passwordConfirm"
                                onChange={[Function]}
                                placeholder="비밀번호 확인"
                                type="password"
                              >
                                <styled.input
                                  autoComplete="new-password"
                                  name="passwordConfirm"
                                  onChange={[Function]}
                                  placeholder="비밀번호 확인"
                                  type="password"
                                >
                                  <input
                                    autoComplete="new-password"
                                    className="sc-dlfnuX hrdAtQ"
                                    name="passwordConfirm"
                                    onChange={[Function]}
                                    placeholder="비밀번호 확인"
                                    type="password"
                                  />
                                </styled.input>
                              </AuthInput>
                              <CommonButton
                                cyan={true}
                                fullWidth={true}
                                onClick={[Function]}
                                style={
                                  Object {
                                    "marginTop": "1rem",
                                  }
                                }
                              >
                                <styled.button
                                  cyan={true}
                                  fullWidth={true}
                                  onClick={[Function]}
                                  style={
                                    Object {
                                      "marginTop": "1rem",
                                    }
                                  }
                                >
                                  <button
                                    className="sc-jSgvzq byYyrh"
                                    onClick={[Function]}
                                    style={
                                      Object {
                                        "marginTop": "1rem",
                                      }
                                    }
                                  >
                                    회원가입
                                  </button>
                                </styled.button>
                              </CommonButton>
                            </form>
                            <AuthFooter>
                              <styled.div>
                                <div
                                  className="sc-eCstlR kpGOUC"
                                >
                                  <Link
                                    to="/login"
                                  >
                                    <LinkAnchor
                                      href="/login"
                                      navigate={[Function]}
                                    >
                                      <a
                                        href="/login"
                                        onClick={[Function]}
                                      >
                                        로그인
                                      </a>
                                    </LinkAnchor>
                                  </Link>
                                </div>
                              </styled.div>
                            </AuthFooter>
                          </AuthForm>
                        </div>
                      </styled.div>
                    </AuthArea>
                  </withRouter(AuthArea)>
                </div>
              </styled.div>
            </div>
          </styled.div>
        </AuthTemplate>
      </RegisterPage>
    </Router>
  </BrowserRouter>
</Provider>
`;
